### DISCUSSION TITLE
[Proposal]: Set default accessability for property and event

### DISCUSSION THREAD

**Discussion by dczuo** [dotnet/csharplang#4509] [Language Ideas] [answered]
The default accessability for class members is now private, which prevents careless opening of private members. The default private in practice applies to fields, but events and properties are mostly public.
Please provide the \<PropertyDefaultAccessability\>public\<PropertyDefaultAccessability\> and \<EventDefaultAccessability\>public\<EventDefaultAccessability\> in the proj file, that can make code such as plain object and DTO purer, for example
```
public class Product
{
    string Name {get; set;}
    string Code {get; set;}
    decimal UnitPrice {get; set;}
}
```

**Comment by YairHalberstadt** [comment, marked as answer]
Currently there are no dialects of C#, where an external configuration variable changes the semantics of a piece of code. This is because dialects have enormous costs, both to the ecosystem, and in complexity of the language/compiler.

In order for something to change this it would have to have absolutely overwhelming benefits.

**Reply by jnm2** [reply]
I think something similar was considered and rejected as part of the records feature, where `record X { string Y; }` would be a way to declare a public property. Am I remembering correctly? Maybe there are notes explaining why they didn't like this that would be relevant to this discussion.

**Reply by 333fred** [reply]
You are remembering correctly. We decided that it added to much in terms of complexity to the language. We'd need a way to make them a field instead, which would be entirely new, and we thought that it would be very confusing. The `data` keyword was considered to avoid that (makes it automatically a public init-only prop), but we haven't really made any progress on designing that for real.

**Reply by dczuo** [reply]
> Currently there are no dialects of C#, where an external configuration variable changes the semantics of a piece of code. This is because dialects have enormous costs, both to the ecosystem, and in complexity of the language/compiler.
> 
> In order for something to change this it would have to have absolutely overwhelming benefits.
* The \<PropertyDefaultAccessibility\>public\<PropertyDefaultAccessibility\> setting is optional and does not change the original behavior, use only when determining suitability, does not affect compilation results, and is not considered to have an impact on the ecosystem.
* It is only when accessibility modifiers are missing at the semantic level, first read whether the default accessibility modifiers is configured, the implementation is not considered to be complex.
* Changing semantics is a real problem. But there are many features that make it difficult to find a universal solution, and the compiler provides some options for suitable scenarios to use that are a had to be chosen to move forward.

**Reply by HaloFour** [reply]
@dczuo 

> The <PropertyDefaultAccessibility>public<PropertyDefaultAccessibility> setting is optional

Having a compiler option is what makes this is a dialect.  The team is very opposed to creating new dialects of the language unless the benefit is absolutely massive (e.g. NRTs).  The benefit here is dubious, at best.  At best you save a few keystrokes, most of which will be written by an IDE with auto-complete.  At worst you accidentally expose a member publicly that you never intended to.  IMO, making a member public should be a very explicit decision that is encoded in the source.

**Reply by dczuo** [reply]
> Having a compiler option is what makes this is a dialect
Thanks for your explanation. I now use a customized MSBuild task to regenerate a new file to replace the original file to simplify entity definition.
